From 310bd7aa830bde7f4b186cf4c47425cc0220b411 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Fri, 23 Apr 2021 18:37:19 +0200
Subject: [PATCH] PONRTSYS-8873: Fix meter configuration for > 4GBit/s

The CIR and PIR in the struct dp_meter_cfg are in bit/s and not in
kbit/s. They are divided by 1000 before they are given to
GSW_QoS_meterCfg_t. The GSW_QoS_meterCfg_t expects values in
Kbit/s according to its documentation. The PON QoS driver also expects
them to be in bit/s and this works for lower rates. The maximum
value for a u32 is 4294967296, as soon as we configure a rate over
4.3 GBit/s this 32 bit value will overflow. Use a 64 bit value instead
and update the comments under the assumption that the documentation
of GSW_QoS_meterCfg_t is correct.

This should fix a problem when we configure a rate of 10GBit/s and are
limited to 1.29GBit/s.
---
 drivers/net/datapath/dpm/gswip31/datapath_gswip.c |  4 ++--
 drivers/net/datapath/pon_qos/pon_qos_tc_police.c  |  2 +-
 include/net/datapath_api_qos.h                    | 12 ++++++------
 3 files changed, 9 insertions(+), 9 deletions(-)

diff --git a/drivers/net/datapath/dpm/gswip31/datapath_gswip.c b/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
index 1613abd8e87e..427835e79d01 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
@@ -19,8 +19,8 @@
 	       (br_cfg)->bSubMeteringEnable[(index)] = bflag; \
 	       (br_cfg)->nTrafficSubMeterId[(index)] = id; })
 
-#define METER_CIR(cir)  ((cir) / 1000)
-#define METER_PIR(pir)  ((pir) / 1000)
+#define METER_CIR(cir)  div_u64(cir, 1000)
+#define METER_PIR(pir)  div_u64(pir, 1000)
 
 #define MAX_TC		16
 #define MAX_FLOW	4
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_police.c b/drivers/net/datapath/pon_qos/pon_qos_tc_police.c
index c272d3267919..7673c00a6b45 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_police.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_police.c
@@ -119,7 +119,7 @@ static bool pon_qos_police_parse_actions(struct tc_cls_flower_offload *f,
 			meter_cfg->pbs = tcf_police_tcfp_mtu(a);
 
 			netdev_dbg(pd->dev,
-				   "%s: cir %u pir %u cbs %u pbs %u\n",
+				   "%s: cir %llu pir %llu cbs %u pbs %u\n",
 				   __func__,
 				   meter_cfg->cir,
 				   meter_cfg->pir,
diff --git a/include/net/datapath_api_qos.h b/include/net/datapath_api_qos.h
index 1dee2317436d..c580695aa380 100644
--- a/include/net/datapath_api_qos.h
+++ b/include/net/datapath_api_qos.h
@@ -1177,13 +1177,13 @@ struct dp_meter_cfg {
 	int meter_id;
 	/*!< meter type single/dual rate */
 	enum dp_meter_type type;
-	/*!< Committed information rate in kbit/s */
-	u32 cir;
-	/*!< Peak information rate in kbit/s */
-	u32 pir;
-	/*!< committed burst size in kbit */
+	/*!< Committed information rate in bit/s */
+	u64 cir;
+	/*!< Peak information rate in bit/s */
+	u64 pir;
+	/*!< committed burst size in bytes */
 	u32 cbs;
-	/*!< peak burst size in kbit */
+	/*!< peak burst size in bytes */
 	u32 pbs;
 	/*!< color blind/aware */
 	bool col_mode;
